packages/eslint-plugin/src/rules/strict-void-return.ts

```diff
@@ -0,0 +1,991 @@
+      nonVoidFuncInOther:
+        'Value-returning function used in a context where a void function was expected.',
+
+      suggestWrapInAsyncIIFE:
+        'Wrap the function body in an immediately-invoked async function expression.',
+      suggestWrapInTryCatch: 'Wrap the function body in a try-catch block.',
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          considerOtherOverloads: { type: 'boolean' },
+          considerBaseClass: { type: 'boolean' },
+          considerImplementedInterfaces: { type: 'boolean' },
+          allowReturnPromiseIfTryCatch: { type: 'boolean' },
+          allowReturnUndefined: { type: 'boolean' },
+          allowReturnNull: { type: 'boolean' },
+          allowReturnAny: { type: 'boolean' },
+        },
+        additionalProperties: false,
+      },
+    ],
+  },
+  defaultOptions: [
+    {
+      considerOtherOverloads: true,
+      considerBaseClass: true,
+      considerImplementedInterfaces: true,
+      allowReturnPromiseIfTryCatch: true,
+      allowReturnUndefined: true,
+      allowReturnNull: true,
+    },
+  ],
+
+  create(context, [options]) {
+    const sourceCode = context.sourceCode;
+    const parserServices = util.getParserServices(context);
+    const checker = parserServices.program.getTypeChecker();
+
+    return {
+      'CallExpression, NewExpression': (
+        node: TSESTree.CallExpression | TSESTree.NewExpression,
+      ): void => {
+        checkFunctionCallNode(node);
+      },
+      JSXAttribute: (node): void => {
+        if (
+          node.value?.type === AST_NODE_TYPES.JSXExpressionContainer &&
+          node.value.expression.type !== AST_NODE_TYPES.JSXEmptyExpression
+        ) {
+          const attrName = sourceCode.getText(node.name);
+          const elemName = sourceCode.getText(node.parent.name);
+          checkExpressionNode(node.value.expression, 'Attr', {
+            attrName,
+            elemName,
+          });
+        }
+      },
+      VariableDeclarator: (node): void => {
+        if (node.init != null) {
+          const varName = util.getNameFromExpression(sourceCode, node.id);
+          if (varName != null) {
+            checkExpressionNode(node.init, 'Var', { varName });
+          } else {
+            checkExpressionNode(node.init, 'Other');
+          }
+        }
+      },
+      AssignmentExpression: (node): void => {
+        if (['=', '||=', '&&=', '??='].includes(node.operator)) {
+          const varName = util.getNameFromExpression(sourceCode, node.left);
+          if (varName != null) {
+            checkExpressionNode(node.right, 'Var', { varName });
+          } else {
+            checkExpressionNode(node.right, 'Other');
+          }
+        }
+      },
+      ObjectExpression: (node): void => {
+        for (const propNode of node.properties) {
+          if (propNode.type !== AST_NODE_TYPES.SpreadElement) {
```

Comment:
[Testing] If I remove this `if` and just leave its body, all unit tests still pass. That means there's at least a missing test case.

Reply:
That's because technically this condition is not necessary, because if it's a spread argument then the contextual type wont be a void function. I don't know how to write a test for this. Is it even possible?